dao/dao_client.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package dao
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	sp "cloud.google.com/go/spanner"
+)
+
+var once sync.Once
+var spClient *sp.Client
+
+// This function is declared as a global variable to make it testable. The unit
+// tests edit this function, acting like a double.
+var newClient = sp.NewClient
+
+func GetOrCreateClient(ctx context.Context, dbURI string) (*sp.Client, error) {
+	var err error
```

Comment:
We already have a client implementation that takes in a `dbURI` and creates a client I think, is there need for another? Looks like a duplicate.
https://github.com/GoogleCloudPlatform/spanner-migration-tool/blob/master/accessors/clients/spanner/client/spanner_client.go

Reply:
The spannerclient is for interactions with the targetDatabase of the customer and DAO is for the metadata database. Since we cannot reuse clients as one client is responsible for one db only, we 2 need separate ones for each.

There might also be cases in the future when a flow could need 3 databases: metadatadb, targetDatabase and some other metadatadb used by templates, in this case a new spanner client will need to be created and closed within the scope of the function where it is being called.